### PR TITLE
Add product-tests for bucketed hive tables

### PR DESCRIPTION
**Pull request by kokosing**
Add product-tests for bucketed hive tables

### CODE REVIEW

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java [hunk 1]
```diff
@@ -21,42 +23,70 @@
...
-                    "n_regionkey     BIGINT," +
-                    "n_comment       STRING) " +
-                    "PARTITIONED BY (part_key STRING) " +
-                    "CLUSTERED BY (n_regionkey) " +
-                    "INTO 2 BUCKETS " +
-                    "ROW FORMAT DELIMITED FIELDS TERMINATED BY '|'")
+    public static final HiveTableDefinition BUCKETED_NATION = bucketTableDefinition("bucket_nation", false, false);
+
+    @TableDefinitionsRepository.RepositoryTableDefinition
+    public static final HiveTableDefinition BUCKETED_NATION_PREPATED = HiveTableDefinition.builder("bucket_nation_prepared")
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java]: typo

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java [hunk 2]
```diff
@@ -21,42 +23,70 @@
...
+            .setCreateTableDDLTemplate("Table %NAME% should be only used with CTAS queries")
             .setNoData()
             .build();
 
+    @TableDefinitionsRepository.RepositoryTableDefinition
+    public static final HiveTableDefinition BUCKETED_SORTED_NATION = bucketTableDefinition("bucket_sort_nation", true, false);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java]: bucket_sort_nation -> bucketed_sorted_nation

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java [hunk 3]
```diff
@@ -21,42 +23,70 @@
...
+
+    @TableDefinitionsRepository.RepositoryTableDefinition
+    public static final HiveTableDefinition BUCKETED_PARTITIONED_NATION = bucketTableDefinition("bucket_partition_nation", false, true);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java]: bucket_partition_nation -> bucketed_partitioned_nation

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java [hunk 4]
```diff
@@ -93,6 +123,154 @@ public void testAllowMultipleFilesPerBucket()
                 .containsExactly(row(3));
     }
 
+    @Test
+    public void testSelectStar()
+    {
+        String tableName = mutableTableInstanceOf(BUCKETED_NATION).getNameInDatabase();
+        populateHiveTable(tableName, NATION.getName());
+
+        assertThat(query(format("SELECT * FROM %s", tableName))).matches(PRESTO_NATION_RESULT);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java]: `format` is redundant

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java [hunk 5]
```diff
@@ -93,6 +123,154 @@ public void testAllowMultipleFilesPerBucket()
                 .containsExactly(row(3));
     }
 
+    @Test
+    public void testSelectStar()
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java]: This is apparently the most basic test, move it first.

File: presto-product-tests/src/main/java/io/prestosql/tests/utils/TableDefinitionUtils.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+public class TableDefinitionUtils
+{
+    private TableDefinitionUtils() {}
+
+    public static TableInstance mutableTableInstanceOf(TableDefinition tableDefinition)
+    {
+        if (tableDefinition.getDatabase().isPresent()) {
+            return mutableTableInstanceOf(tableDefinition, tableDefinition.getDatabase().get());
+        }
+        else {
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/utils/TableDefinitionUtils.java]: `else` is redundant

File: presto-product-tests/src/main/java/io/prestosql/tests/utils/TableDefinitionUtils.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+import io.prestosql.tempto.fulfillment.table.MutableTablesState;
+import io.prestosql.tempto.fulfillment.table.TableDefinition;
+import io.prestosql.tempto.fulfillment.table.TableHandle;
+import io.prestosql.tempto.fulfillment.table.TableInstance;
+
+import static io.prestosql.tempto.context.ThreadLocalTestContextHolder.testContext;
+import static io.prestosql.tempto.fulfillment.table.TableHandle.tableHandle;
+
+public class TableDefinitionUtils
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/utils/TableDefinitionUtils.java]: `final`

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java [hunk 8]
```diff
@@ -93,6 +123,154 @@ public void testAllowMultipleFilesPerBucket()
...
+        assertThat(query(format("SELECT * FROM %s", tableName))).matches(PRESTO_NATION_RESULT);
+        assertThat(query(format("SELECT count(*) FROM %s WHERE n_regionkey=0", tableName))).containsExactly(row(5));
+    }
+
+    private static void populateRowToHiveTable(String destination, List<String> values, Optional<String> partition)
+    {
+        String queryStatement = format("INSERT INTO TABLE %s" +
+                        (partition.isPresent() ? format(" PARTITION (%s) ", partition.get()) : " ") +
+                        "SELECT %s from (select 'foo') x",
+                destination, Joiner.on(",").join(values));
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java]: Joiner -> `String.join(..)` (static import)

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java [hunk 9]
```diff
@@ -93,6 +123,154 @@ public void testAllowMultipleFilesPerBucket()
...
+        assertThat(query(format("SELECT count(*) FROM %s WHERE n_regionkey=0 AND part_key='ALGERIA'", tableName))).containsExactly(row(1));
+    }
+
+    @Test
+    public void testInsertIntoBucketedTables()
+    {
+        String tableName = mutableTablesState().get(BUCKETED_NATION).getNameInDatabase();
+
+        assertThat(() -> query(format("INSERT INTO %s SELECT * FROM %s", tableName, NATION.getName())))
+                .failsWithMessage("Cannot insert into bucketed unpartitioned Hive table");
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveBucketedTables.java]: Is it true now that https://github.com/prestosql/presto/pull/924 is merged?
Please rebase to force Travis re-merge anew.

**Comment by kokosing** [comment]
Tests failure is unrelated. Merging.

**Pull request merged**